dpctl/tensor/_ctors.py

```diff
@@ -763,6 +763,8 @@ def full(
         order = order[0].upper()
     dpctl.utils.validate_usm_type(usm_type, allow_none=False)
     sycl_queue = normalize_queue_device(sycl_queue=sycl_queue, device=device)
+    if dtype is None and isinstance(fill_value, (dpt.usm_ndarray, np.ndarray)):
+        dtype = fill_value.dtype
```

Comment:
This is not enough, I am afraid. If `fill_value` is `np.ndarray`, we need to ensure it has just 1 element, to extract the scalar from it, and get `dtype` like you are doing. 

If `fill_value` is `usm_ndarray` we need to check that compute-follows-data rules apply, then broadcast `fill_value` to `sh`, and apply copy-with-casting operator.

All other cases must proceed as before.